docs/man/man3/main_parser.3
.TH "main_parser" 3 "Thu May 6 2021" "My Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
main_parser
.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "def \fBassessmentMethodProcessor\fP (assessment_methods)"
.br
.ti -1c
.RI "def \fBregex_inc\fP (regex, match_list, header_text)"
.br
.RI "Applies a regular expression to a block of text\&. "
.ti -1c
.RI "def \fBtable_two_access\fP (is_accredited, column, table, row_iter, data, one_cols, max_rows)"
.br
.RI "Captures the information in the Assessment Methods table(s) of the document\&. "
.ti -1c
.RI "def \fBgrad_unaccredited_assessment\fP (column, report_year, doc, slo_count, data, one_cols, max_rows)"
.br
.RI "Captures the information in Non-Accredited Graduate reports in the Assessment Methods table(s) of the document\&. "
.ti -1c
.RI "def \fBtable_three_access\fP (column, table, row_iter, data, one_cols, max_rows, slo_count)"
.br
.RI "Captures the information in the Data (Collection &) Analysis table of the document\&. "
.ti -1c
.RI "def \fBtable_four_access\fP (is_accredited, column, table, row_iter, data_coll_list, slo_count)"
.br
.RI "Captures the information in the Decisions & Actions table of the document\&. "
.ti -1c
.RI "def \fBrun\fP (uploaded_filename)"
.br
.RI "This function captures the information from documents\&. "
.in -1c
.SH "Function Documentation"
.PP 
.SS "def main_parser\&.assessmentMethodProcessor ( assessment_methods)"

.SS "def main_parser\&.grad_unaccredited_assessment ( column,  report_year,  doc,  slo_count,  data,  one_cols,  max_rows)"

.PP
Captures the information in Non-Accredited Graduate reports in the Assessment Methods table(s) of the document\&. 
.PP
\fBParameters\fP
.RS 4
\fIcolumn\fP - A tuple that refers to the column number we are accessing in the table\&. 
.br
\fIreport_year\fP - The year the report was created\&. 
.br
\fIdoc\fP - The document object, so that we can iterate through its tables\&. 
.br
\fIdata\fP - A list that the captured data is appended to\&. 
.br
\fIone_cols\fP - The assessment methods tables have three 'columns' that are just rows\&. This is represented as 2 for 0-based arrays\&. 
.br
\fImax_rows\fP - The maximum number of rows that should be in the table we are accessing\&.
.RE
.PP
\fBReturns\fP
.RS 4
The captured data in a list\&. 
.RE
.PP

.SS "def main_parser\&.regex_inc ( regex,  match_list,  header_text)"

.PP
Applies a regular expression to a block of text\&. 
.PP
\fBParameters\fP
.RS 4
\fIregex\fP - This is the regular expressions to apply\&. 
.br
\fImatch_list\fP - This list keeps track of all the matches that the regular expressions have caught\&. 
.br
\fIheader_text\fP - This is the text that the regular expression is compared to\&.
.RE
.PP
\fBReturns\fP
.RS 4
The list of matches\&. 
.RE
.PP

.SS "def main_parser\&.run ( uploaded_filename)"

.PP
This function captures the information from documents\&. 
.PP
\fBParameters\fP
.RS 4
\fIuploaded_filename\fP This is the file name of the document we are running the method on\&. 
.RE
.PP

.SS "def main_parser\&.table_four_access ( is_accredited,  column,  table,  row_iter,  data_coll_list,  slo_count)"

.PP
Captures the information in the Decisions & Actions table of the document\&. 
.PP
\fBParameters\fP
.RS 4
\fIis_accredited\fP - Boolean value, True if the report is for an accredited degree program\&. 
.br
\fIcolumn\fP - A tuple that refers to the column number we are accessing in the table\&. 
.br
\fItable\fP - The docx\&.Table object that we are accessing\&. 
.br
\fIrow_iter\fP - An integer that keeps track of what row we are accessing\&. 
.br
\fIdata_coll_list\fP - A list that the captured data is appended to\&. 
.br
\fIone_cols\fP - I honestly forgot how I was using this variable\&. 
.br
\fIslo_count\fP - This integer is the number of SLOs that were counted in the document\&.
.RE
.PP
\fBReturns\fP
.RS 4
The captured data in a list of tuples\&. 
.RE
.PP

.SS "def main_parser\&.table_three_access ( column,  table,  row_iter,  data,  one_cols,  max_rows,  slo_count)"

.PP
Captures the information in the Data (Collection &) Analysis table of the document\&. 
.PP
\fBParameters\fP
.RS 4
\fIcolumn\fP - A tuple that refers to the column number we are accessing in the table\&. 
.br
\fItable\fP - The docx\&.Table object that we are accessing\&. 
.br
\fIrow_iter\fP - An integer that keeps track of what row we are accessing\&. 
.br
\fIdata\fP - A list that the captured data is appended to\&. 
.br
\fIone_cols\fP - I honestly forgot how I was using this variable\&. 
.br
\fImax_rows\fP - The maximum number of rows that should be in the table we are accessing\&. 
.br
\fIslo_count\fP - This integer is the number of SLOs that were counted in the document\&.
.RE
.PP
\fBReturns\fP
.RS 4
The captured data in a list\&. 
.RE
.PP

.SS "def main_parser\&.table_two_access ( is_accredited,  column,  table,  row_iter,  data,  one_cols,  max_rows)"

.PP
Captures the information in the Assessment Methods table(s) of the document\&. 
.PP
\fBParameters\fP
.RS 4
\fIis_accredited\fP - Boolean value, True if the report is for an accredited degree program\&. 
.br
\fIcolumn\fP - A tuple that refers to the column number we are accessing in the table\&. 
.br
\fItable\fP - The docx\&.Table object that we are accessing\&. 
.br
\fIrow_iter\fP - An integer that keeps track of what row we are accessing\&. 
.br
\fIdata\fP - A list that the captured data is appended to\&. 
.br
\fIone_cols\fP - The assessment methods tables have three 'columns' that are just rows\&. This is represented as 2 for 0-based arrays\&. 
.br
\fImax_rows\fP - The maximum number of rows that should be in the table we are accessing\&.
.RE
.PP
\fBReturns\fP
.RS 4
The captured data in a list\&. 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for My Project from the source code\&.


mysite/core/paser/main_parser.py
import docx
from docx.oxml.ns import qn
import re
from mysite.core.paser import read_doc_chx
import psycopg2
from mysite.core.paser import insertData

def assessmentMethodProcessor(assessment_methods):
    output = []
    temp = []
    for item in assessment_methods:
        if (type(item) == str and item.startswith("SLO")):
            if (len(temp) != 0):
                output.append(temp)
                temp = []
        temp.append(item)
    output.append(temp)
    return output

def regex_inc(regex, match_list, header_text):
    """!  Applies a regular expression to a block of text.
    
    @param regex - This is the regular expressions to apply.
    @param match_list - This list keeps track of all the matches that the regular expressions have caught.
    @param header_text - This is the text that the regular expression is compared to.

    @return The list of matches.
    """
    rex = re.findall(regex, header_text)
    for match in rex[0]:
        match_list.append(match)
    return (match_list)

def table_two_access(is_accredited, column, table, row_iter, data, one_cols, max_rows):
    """!  Captures the information in the Assessment Methods table(s) of the document.
    
    @param is_accredited - Boolean value, True if the report is for an accredited degree program.
    @param column - A tuple that refers to the column number we are accessing in the table.
    @param table - The docx.Table object that we are accessing.
    @param row_iter - An integer that keeps track of what row we are accessing.
    @param data - A list that the captured data is appended to.
    @param one_cols - The assessment methods tables have three "columns" that are just rows. This is represented as 2 for 0-based arrays.
    @param max_rows - The maximum number of rows that should be in the table we are accessing.

    @return The captured data in a list.
    """    
    if (not is_accredited):
        for row in table.rows:
            if row_iter <= one_cols or (row_iter > max_rows and row_iter <= one_cols + max_rows):
                data.append(row.cells[column[1]].text)
            if (row_iter > one_cols and row_iter < max_rows + 1) or (row_iter > max_rows + one_cols):
                data.append((row.cells[0].text, row.cells[column[2]].text))
            row_iter += 1
            if row_iter > max_rows:
                row_iter = 0
    else:
        for row in table.rows:
            if (row_iter == 0):
                row_iter += 1
                continue
            data.append(("\n\n" + "SLO " + row.cells[column[1]].text, row.cells[column[2]].text, row.cells[column[3]].text, row.cells[column[4]].text))
    return data

def grad_unaccredited_assessment(column, report_year, doc, slo_count, data, one_cols, max_rows):
    """!  Captures the information in Non-Accredited Graduate reports in the Assessment Methods table(s) of the document.
    
    @param column - A tuple that refers to the column number we are accessing in the table.
    @param report_year - The year the report was created.
    @param doc - The document object, so that we can iterate through its tables.
    @param data - A list that the captured data is appended to.
    @param one_cols - The assessment methods tables have three "columns" that are just rows. This is represented as 2 for 0-based arrays.
    @param max_rows - The maximum number of rows that should be in the table we are accessing.

    @return The captured data in a list.
    """ 
    row_iter = 0
    slo_iter = 0
    table = doc.tables[1]   
    if (report_year == 18):
        for row in table.rows:
            ## If we are still on one of the first three columns of the table, or we have gone through the first table and onto the second.
            ## Only append the first column, because there is only one column.
            if row_iter <= one_cols or (row_iter > max_rows and row_iter <= one_cols + max_rows):
                data.append(row.cells[column[1]].text)
            ## If we are past the first three columns and have not moved on to another table, append the first and second column data.
            if (row_iter > one_cols and row_iter < max_rows + 1) or (row_iter > max_rows + one_cols):
                data.append((row.cells[0].text, row.cells[column[2]].text))
        table = doc.tables[3]
        for row in table.rows:
            ## If we are still on one of the first three columns of the table, or we have gone through the first table and onto the second.
            ## Only append the first column, because there is only one column.
            if row_iter <= one_cols or (row_iter > max_rows and row_iter <= one_cols + max_rows):
                data.append(row.cells[column[1]].text)
            ## If we are past the first three columns and have not moved on to another table, append the first and second column data.
            if (row_iter > one_cols and row_iter < max_rows + 1) or (row_iter > max_rows + one_cols):
                data.append((row.cells[0].text, row.cells[column[2]].text))
    return data


def table_three_access(column, table, row_iter, data, one_cols, max_rows, slo_count):
    """!  Captures the information in the Data (Collection &) Analysis table of the document.
    
    @param column - A tuple that refers to the column number we are accessing in the table.
    @param table - The docx.Table object that we are accessing.
    @param row_iter - An integer that keeps track of what row we are accessing.
    @param data - A list that the captured data is appended to.
    @param one_cols - I honestly forgot how I was using this variable.
    @param max_rows - The maximum number of rows that should be in the table we are accessing.
    @param slo_count - This integer is the number of SLOs that were counted in the document.

    @return The captured data in a list.
    """    

    ### Original loop from link below
    ### https://www.reddit.com/r/learnpython/comments/dbie6s/help_iterating_over_a_table_in_pythondocx/
    for row in table.rows:
        if row.cells[column[0]].text == "SLO 1":
            break
        if row_iter <= one_cols or (row_iter > max_rows and row_iter <= one_cols + max_rows):
            data.append(row.cells[column[1]].text)
        if (row_iter > one_cols and row_iter < max_rows + 1) or (row_iter > max_rows + one_cols):
            data.append((row.cells[0].text, row.cells[column[2]].text))
        row_iter += 1
        if row_iter > max_rows:
            row_iter = 0
    return data

def table_four_access(is_accredited, column, table, row_iter, data_coll_list, slo_count):
    """!  Captures the information in the Decisions & Actions table of the document.
    
    @param is_accredited - Boolean value, True if the report is for an accredited degree program.
    @param column - A tuple that refers to the column number we are accessing in the table.
    @param table - The docx.Table object that we are accessing.
    @param row_iter - An integer that keeps track of what row we are accessing.
    @param data_coll_list - A list that the captured data is appended to.
    @param one_cols - I honestly forgot how I was using this variable.
    @param slo_count - This integer is the number of SLOs that were counted in the document.

    @return The captured data in a list of tuples.
    """
    
    if (not is_accredited):
        for row in table.rows:
            if "SLO " + str(slo_count + 1) in row.cells[column[1]].text:
                break
            slo_measure = row.cells[column[1]].text
            data_coll_range = row.cells[column[2]].text
            num_students = row.cells[column[3]].text
            percent_students = row.cells[column[4]].text
            extend_values = [(slo_measure, data_coll_range, num_students, percent_students)]
            if data_coll_range == "" and num_students == "" and percent_students == "":
                continue
            data_coll_list.extend(extend_values)
    else:
        for row in table.rows:
            slo_num = row.cells[column[1]].text
            dec_and_act = row.cells[column[2]].text
            extend_values = [(slo_num, dec_and_act)]
            data_coll_list.extend(extend_values)
            row_iter += 1
    return data_coll_list

def run(uploaded_filename):
    """! This function captures the information from documents.

    @param uploaded_filename This is the file name of the document we are running the method on.
    """

    ## Whether the report is undergraduate; False = graduate, True = undergraduate.
    is_undergrad = False
    ## Whether the report is accredited; False = Non-accredited, True = accredited 
    is_accredited = False

    ## Report year, either 0 or 19
    report_year = 0

    ## The regular expression used to capture the "B" or "M" for Bachelors or Masters, in the file name. "B" = undergraduate, anything else = graduate.
    filename_regex = "\.*?\((.{1})\w+\)\.*"

    ## The match object returned when checking the uploaded file name against the file name regular expression defined above.
    filename_match = re.search(filename_regex, uploaded_filename)

    if (filename_match.group(1).lower() == "b"):
        is_undergrad = True

    ## The name we use to create the ZIP version of the document. It should be "filename".zip.
    unzip_name = uploaded_filename[:-5] + ".zip"
    ## The directory that we use to store the documents after unzipping the ZIP file.
    zip_dir = uploaded_filename[:-5] + "/"
    ## The file directory that we store all the documents.
    file_dir = "./media/"
    ## A tuple that stores the table numbers where the document information is stored.
    table_access_list = ()

    ## The docx object that we use to access the document information gained from the docx module.
    doc = docx.Document(file_dir + uploaded_filename)

    ## Counter to keep track of how many iterations through the paragraphs we have gone.
    para_iter = 0

    ## Loop through all paragraphs until we find whether it is Non-Accredited or Accredited
    for para in doc.paragraphs:
        if (para_iter > 10):
            break
        if ("Non".lower() in para.text.lower()):
            break
        elif ("Accredited".lower() in para.text.lower()):
            is_accredited = True
            break
        para_iter += 1

    read_doc_chx.copy_and_unzip(file_dir, uploaded_filename, unzip_name, zip_dir)

    ## Determines the appropriate regular expression list for the document header and table access based on undergraduate/graduate, accredited/non, and report year.
    if (not is_accredited):
        header_regex = 'College:\s*(.*)\s*Department\/School:\s*(.*)\s*Program:\s*(.*)\s*Degree Level:\s*(.*)\s*Academic Year of Report:\s*(.*)\s*Date Range of Reported Data:\s*(.*)\s*Person Preparing the Report:\s*(.*)'
        report_header_list = ['College: ', 'Department/School: ', 'Program: ', 'Degree Level: ', 'Academic Year of Report: ', 'Date: ', 'Person Preparing the Report: ']
    else:
        header_regex = 'College:\s*(.*)\s*Department/School:\s*(.*)\s*Program:\s*(.*)\s*Degree Level:\s*(.*)\s*Academic Year of Report:\s*(.*)\s*Person Preparing the Report:\s(.*)\s*Last Accreditation Review:\s(.*)\s*Accreditation Body:\s(.*)\s{2}'
        report_header_list = ['College: ', 'Department/School: ', 'Program: ', 'Degree Level: ', 'Academic Year of Report', 'Person Preparing the Report: ', 'Last Accreditation Review: ', 'Accreditation Body: ']
        table_access_list = (0, 1, 2, 3)

    ## The path in the unzipped directory to the document XML file.
    xml_string = "word/document.xml"
    ## The full path to the unzipped directory.
    xml_path = file_dir + zip_dir + xml_string

    #assessment_obj = assessment_models.Assessment()
    
    ## Store the return values from the find_checkbox_elements function.
    (chkbox_element_list, slo_count, list_of_lists) = read_doc_chx.find_checkbox_elements(xml_path)

    ## Printing out the checkbox elements, ONLY USED FOR TESTING PURPOSES.
    print("Checkbox elements here: printing at Line 209")
    print("============================================================================================")

    for item in list_of_lists:
        print(item)
    print("============================================================================================\n")

    ## A counter to keep track of what regular expression in a list we are on.
    regex_counter = 0
    
    ## A list object used to hold the matches found from the regular expressions.
    match_list = []

    ## Stores all the docx Paragraph objects found in the document.
    all_paras = doc.paragraphs

    para_iter = 0
    header_text = ""
    ## Loop through all of the paragraphs and compare them to the regular expressions.
    for para in all_paras:
        header_text += "\n " + para.text
        if (para_iter < 12):
            para_iter += 1
        else:
            break

    print(header_text)
    
    match_list = regex_inc(header_regex, match_list, header_text)
    print(match_list)
    ## Print out the report header info, ONLY FOR TESTING PURPOSES.
    print("Report Header info: printing at line 231")
    print("============================================================================================")
    
    for i in range(0, len(report_header_list)):
        print(report_header_list[i] + ": " + match_list[i])
    print("============================================================================================\n")

    master_table = (0, 1, 2)
    if (not is_accredited and "18" in match_list[4]):
        report_year = 18
        table_access_list = master_table + (3, 5)
    elif (not is_accredited and is_undergrad):
        table_access_list = master_table + (4, 6)
    elif (is_accredited):
        table_access_list = master_table + (3,)
    else:
        report_year = 19

    ## List used to store information captured from tables.
    data = []

    ## List used to store SLOs from document.
    slo_list = []

    ### Original code from https://stackoverflow.com/questions/27861732/parsing-of-table-from-docx-file/27862205 ###
    ## Used to store the SLO Table information that we are interested in.
    table = doc.tables[table_access_list[0]]
    ## Append the header rows as keys, and the cells beneath them as the corresponding values
    for i, row in enumerate(table.rows):
        text = (cell.text for cell in row.cells)
        
        ## Establish the mapping based on the first row headers; these will become the keys of our dictionary.
        if i == 0:
            keys = tuple(text)
            continue
        elif i == 2 * (slo_count - 1):
            break

        ## Construct a dictionary for this row, mapping keys to values for this row.
        row_data = dict(zip(keys, text))
        data.append(row_data)

    ## Append the SLOs to the SLO list
    for slo in range (0, slo_count):
        if (not is_undergrad and report_year == 19):
            slo_list.append(data[slo]['\nStudent Learning Outcomes'])
        else:
            slo_list.append(data[slo]['Student Learning Outcomes'])

    ## Clear data so that we can use it again.
    data.clear()

    ## Print out the list of SLOs, ONLY FOR TESTING PURPOSES.
    print("SLO List: printing at line 272")
    print("============================================================================================")
    for slo in slo_list:
        print(slo)
    print("============================================================================================\n")

    ## Tuple to hold the 0-based number for each column in a table.
    column = (0, 0, 1, 2, 3)
    ## Integer used to keep track of what row we are accessing on a table.
    row_iter = 0
    ## I still don't remember how I created the logic for this, sorry.
    one_cols = 2
    ## The max rows that should be in the tables we are accessing. The XML for these tables is shot to hell, so we had to create a variable to tell the loop to end.
    max_rows = 10
    ## Set the table variable to the next table we need to access, via the table_access_list.
    table = doc.tables[table_access_list[1]]
    ## Reset the previous text. Maybe obsolete.
    prev_text = ""
    if (not is_undergrad and not is_accredited):
        data = grad_unaccredited_assessment(column, report_year, doc, slo_count, data, one_cols, max_rows)
    else:
        data = table_two_access(is_accredited, column, table, row_iter, data, one_cols, max_rows)

    ## Reset row_iter, this might not be needed if the row_iter variable is not changed after calling table_two_access(...).
    row_iter = 0
    ## Maybe obsolete.
    prev_text = ""
    ## Set the table to the next table in the table_access_list.
    table = doc.tables[table_access_list[2]]
    assessment_methods = table_three_access(column, table, row_iter, data, one_cols, max_rows, slo_count)

    #added in splitting assessment_methods into one list per SLO
    assessment_methods = assessmentMethodProcessor(assessment_methods)

    ## Print out the Assessment Methods info, ONLY FOR TESTING PURPOSES.
    print("Assessment Methods info: printing at line 304")
    print("============================================================================================")
    for method_info in assessment_methods:
        print(method_info)
    print("============================================================================================\n")

    #### DATA COLLECTION AND ANALYSIS
    ## The list that holds the information from the Data Collection & Analysis table.
    data_coll_list = []
    ## Set the table variable to the next table.
    if (not is_undergrad and not is_accredited):
        table = doc.tables[table_access_list[3] + 1]
    else:
        table = doc.tables[table_access_list[3]]
    data_coll_list = table_four_access(is_accredited, column, table, row_iter, data_coll_list, slo_count)

    ## Print out the Data Collection & Analysis info, ONLY FOR TESTING purposes.
    skip = 0
    print("Data Collection & Analysis info: printing at line 319")
    print("============================================================================================")
    if (is_accredited):
        for data_coll in data_coll_list:
            print(data_coll[0] + " : " + data_coll[1])
    else:
        for data_coll in data_coll_list:
            if skip == 0:
                skip = 1
                continue
            print(data_coll)
    print("============================================================================================\n")

    #### DECISIONS AND ACTIONS
    ## This is only for Non-Accredited reports. 
    if (not is_accredited):
        table = doc.tables[table_access_list[4]]
        dec_act = []
        for row in table.rows:
            slo_num = row.cells[column[1]].text
            slo_act = row.cells[column[2]].text
            if slo_act == "\n\n\n\n" or slo_act == "":
                continue
            dec_act.extend([(slo_num, slo_act)])

        ## Print out Decisions & Actions info, ONLY FOR TESTING PURPOSES.
        print("Decisions & Actions info: printing at line 345")
        print("============================================================================================")
        for info in dec_act:
            print(info[0] + info[1])
        print("============================================================================================\n")
       # insertData.insertCheckBox(list_of_lists)
       # print(len(slo_list))
        insertData.insertReportHeader(match_list, list_of_lists,is_accredited,len(slo_list), dec_act, assessment_methods, slo_list, data_coll_list)